File: snugfit-integration/android/SnugFitBridge.kt
/**
 * ─────────────────────────────────────────────────────────────────────────────
 * SnugFitBridge.kt  –  Android Native Module für React Native
 * ─────────────────────────────────────────────────────────────────────────────
 * Verbindet das React Native JavaScript-Layer mit dem nativen SnugFit SDK.
 *
 * Setup:
 *   1. snugfit-sdk-3.x.aar in android/app/libs/ kopieren
 *   2. build.gradle: implementation(files('libs/snugfit-sdk-3.x.aar'))
 *   3. Berechtigungen in AndroidManifest.xml:
 *      <uses-permission android:name="android.permission.CAMERA"/>
 *      <uses-feature android:name="android.hardware.camera.ar" android:required="false"/>
 *
 * Benötigt Android 9+ (API 28), empfohlen Android 11+ für ToF-Support
 * ─────────────────────────────────────────────────────────────────────────────
 */

package com.atelier.snugfit

import android.app.Activity
import android.view.ViewGroup
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.snugfit.sdk.*          // ← offizielles SnugFit Package (Pseudocode)
import com.snugfit.sdk.ar.*
import com.snugfit.sdk.mesh.*
import com.snugfit.sdk.export.*

class SnugFitBridge(private val reactContext: ReactApplicationContext)
    : ReactContextBaseJavaModule(reactContext), SnugFitEventListener {

  // ── Interne Zustandsvariablen ──────────────────────────────────────────────
  private val activeSessions = mutableMapOf<String, SFScanSession>()
  private val arControllers  = mutableMapOf<String, SFARController>()

  override fun getName() = "SnugFitSDK"

  // ─────────────────────────────────────────────────────────────────────────
  // SDK Initialisierung
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Initialisiert das SnugFit SDK.
   * PSEUDOCODE: SnugFitManager ist der zentrale SDK-Einstiegspunkt.
   */
  @ReactMethod
  fun initialize(config: ReadableMap, promise: Promise) {
    try {
      // PSEUDOCODE ─ SDK-Konfiguration
      val sdkConfig = SFConfiguration.Builder()
        .apiKey(config.getString("apiKey") ?: "")
        .environment(
          if (config.getString("environment") == "sandbox")
            SFEnvironment.SANDBOX else SFEnvironment.PRODUCTION
        )
        .resolution(parseResolution(config.getString("resolution")))
        .meshModel(config.getString("meshModel") ?: "foot_v3")
        .enableARCore(config.getBoolean("enableARMode") ?: true)
        .enableToF(config.getBoolean("enableLiDAR") ?: true)
        .build()

      // PSEUDOCODE ─ SDK initialisieren
      SnugFitManager.getInstance().configure(reactContext, sdkConfig)
      SnugFitManager.getInstance().addEventListener(this)

      val resultMap = Arguments.createMap().apply {
        putString("version",  SnugFitManager.getInstance().sdkVersion)
        putBoolean("toFAvailable", SFDeviceCapabilities.hasToFSensor(reactContext))
        putString("status",   "initialized")
      }
      promise.resolve(resultMap)

    } catch (e: Exception) {
      promise.reject("SF_INIT_ERROR", "SDK-Initialisierung fehlgeschlagen: ${e.message}", e)
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Session Management
  // ─────────────────────────────────────────────────────────────────────────

  @ReactMethod
  fun createScanSession(options: ReadableMap, promise: Promise) {
    val footSideStr = options.getString("footSide") ?: run {
      promise.reject("SF_INVALID_PARAMS", "footSide erforderlich")
      return
    }

    val footSide = when (footSideStr) {
      "LEFT"  -> SFFootSide.LEFT
      "RIGHT" -> SFFootSide.RIGHT
      else    -> { promise.reject("SF_INVALID_PARAMS", "Ungültige footSide: $footSideStr"); return }
    }

    // PSEUDOCODE ─ Session-Optionen konfigurieren
    val sessionOptions = SFSessionOptions.Builder()
      .footSide(footSide)
      .captureTimeout(options.getDouble("captureTimeout").toLong().coerceAtLeast(10L))
      .minQuality(options.getDouble("minQuality").toFloat().coerceIn(0f, 100f))
      .mode(if (options.getString("mode") == "freeform") SFScanMode.FREEFORM else SFScanMode.GUIDED)
      .build()

    val session   = SnugFitManager.getInstance().createSession(sessionOptions)
    val sessionId = session.sessionId

    activeSessions[sessionId] = session

    val resultMap = Arguments.createMap().apply {
      putString("sessionId", sessionId)
    }
    promise.resolve(resultMap)
  }

  // ─────────────────────────────────────────────────────────────────────────
  // AR-Kamera-Preview
  // ─────────────────────────────────────────────────────────────────────────

  @ReactMethod
  fun startARPreview(options: ReadableMap, promise: Promise) {
    val sessionId  = options.getString("sessionId") ?: run {
      promise.reject("SF_PREVIEW_ERROR", "sessionId fehlt")
      return
    }
    val session    = activeSessions[sessionId] ?: run {
      promise.reject("SF_PREVIEW_ERROR", "Session nicht gefunden")
      return
    }
    val viewHandle = options.getInt("viewHandle")

    reactContext.runOnUiQueueThread {
      // PSEUDOCODE ─ Native View aus React Native UIManager holen
      val uiManager  = reactContext.getNativeModule(UIManagerModule::class.java)
      val targetView = uiManager?.resolveView(viewHandle) as? ViewGroup ?: run {
        promise.reject("SF_PREVIEW_ERROR", "View nicht gefunden")
        return@runOnUiQueueThread
      }

      // PSEUDOCODE ─ AR-Controller erstellen
      val arConfig = SFARConfig.Builder()
        .showDepthMap(options.getBoolean("showDepthMap") ?: true)
        .showSkeleton(options.getBoolean("showSkeleton") ?: true)
        .overlayColor(0xFF10B981.toInt())   // Grün
        .build()

      val arController = SnugFitManager.getInstance().createARController(session, arConfig)

      // AR-View in den React Native Container einbetten
      val arView = arController.getView(reactContext)
      val lp     = ViewGroup.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.MATCH_PARENT
      )
      targetView.addView(arView, lp)

      // Position-Updates weiterleiten
      arController.setPositionUpdateListener { positionData ->
        val eventMap = Arguments.createMap().apply {
          putString("sessionId",  sessionId)
          putString("status",     positionData.statusString)
          putDouble("confidence", positionData.confidence.toDouble())
          putMap("boundingBox", Arguments.createMap().apply {
            putDouble("x",      positionData.boundingBox.left.toDouble())
            putDouble("y",      positionData.boundingBox.top.toDouble())
            putDouble("width",  positionData.boundingBox.width().toDouble())
            putDouble("height", positionData.boundingBox.height().toDouble())
          })
        }
        sendEvent("SnugFitPositionUpdate", eventMap)
      }

      arControllers[sessionId] = arController
      arController.start(currentActivity!!)

      val resultMap = Arguments.createMap().apply {
        putString("status", "ar_preview_started")
      }
      promise.resolve(resultMap)
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Winkel-Aufnahme
  // ─────────────────────────────────────────────────────────────────────────

  @ReactMethod
  fun captureAngle(options: ReadableMap, promise: Promise) {
    val sessionId = options.getString("sessionId") ?: run {
      promise.reject("SF_CAPTURE_ERROR", "sessionId fehlt")
      return
    }
    val session  = activeSessions[sessionId] ?: run {
      promise.reject("SF_CAPTURE_ERROR", "Session nicht gefunden")
      return
    }
    val angleId  = options.getString("angleId") ?: "top"

    val captureOptions = SFCaptureOptions.Builder()
      .angleId(angleId)
      .frameCount(options.getInt("frameCount").coerceAtLeast(1))
      .depthResolution(SFDepthResolution.FULL)
      .build()

    // PSEUDOCODE ─ Aufnahme asynchron durchführen
    session.captureAngle(captureOptions, object : SFCaptureCallback {
      override fun onSuccess(result: SFCaptureResult) {
        val resultMap = Arguments.createMap().apply {
          putBoolean("success",  true)
          putDouble("quality",   result.quality.toDouble())
          putString("message",   "Aufnahme erfolgreich")
          putString("previewImageBase64", result.previewImageBase64 ?: "")
        }
        promise.resolve(resultMap)
      }

      override fun onQualityInsufficient(quality: Float, message: String) {
        val resultMap = Arguments.createMap().apply {
          putBoolean("success", false)
          putDouble("quality",  quality.toDouble())
          putString("message",  message)
        }
        promise.resolve(resultMap)   // resolve (kein reject) – UI handelt den Retry
      }

      override fun onError(error: SFException) {
        promise.reject("SF_CAPTURE_ERROR", error.message, error)
      }
    })
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Mesh-Verarbeitung
  // ─────────────────────────────────────────────────────────────────────────

  @ReactMethod
  fun processMesh(options: ReadableMap, promise: Promise) {
    val sessionId = options.getString("sessionId") ?: run {
      promise.reject("SF_MESH_ERROR", "sessionId fehlt")
      return
    }
    val session = activeSessions[sessionId] ?: run {
      promise.reject("SF_MESH_ERROR", "Session nicht gefunden")
      return
    }

    val meshOptions = SFMeshOptions.Builder()
      .targetTriangles(options.getInt("targetTriangles").coerceAtLeast(1_000))
      .smoothing(options.getDouble("smoothing").toFloat().coerceIn(0f, 1f))
      .fillHoles(options.getBoolean("fillHoles") ?: true)
      .coordinateSystem(SFCoordinateSystem.FOOT_STANDARD)
      .build()

    session.processMesh(meshOptions,
      progressCallback = { percentage ->
        val progressMap = Arguments.createMap().apply {
          putString("sessionId",  sessionId)
          putInt("percentage",    percentage)
        }
        sendEvent("SnugFitMeshProgress", progressMap)
      },
      callback = object : SFMeshCallback {
        override fun onSuccess(result: SFMeshResult) {
          val measurementsMap = Arguments.createMap().apply {
            putDouble("length",       result.measurements.length.toDouble())
            putDouble("width",        result.measurements.width.toDouble())
            putDouble("heelWidth",    result.measurements.heelWidth.toDouble())
            putDouble("instepHeight", result.measurements.instepHeight.toDouble())
            putDouble("archHeight",   result.measurements.archHeight.toDouble())
            putDouble("ballGirth",    result.measurements.ballGirth.toDouble())
          }
          val resultMap = Arguments.createMap().apply {
            putDouble("accuracy",     result.accuracy.toDouble())
            putString("meshId",       result.meshId)
            putMap("measurements",    measurementsMap)
          }
          promise.resolve(resultMap)
        }

        override fun onError(error: SFException) {
          promise.reject("SF_MESH_ERROR", "Mesh-Verarbeitung fehlgeschlagen: ${error.message}", error)
        }
      }
    )
  }

  // ─────────────────────────────────────────────────────────────────────────
  // STL-Export
  // ─────────────────────────────────────────────────────────────────────────

  @ReactMethod
  fun exportSTL(options: ReadableMap, promise: Promise) {
    val sessionId  = options.getString("sessionId")  ?: run {
      promise.reject("SF_EXPORT_ERROR", "sessionId fehlt"); return
    }
    val session    = activeSessions[sessionId] ?: run {
      promise.reject("SF_EXPORT_ERROR", "Session nicht gefunden"); return
    }
    val outputPath = options.getString("outputPath") ?: run {
      promise.reject("SF_EXPORT_ERROR", "outputPath fehlt"); return
    }

    val exportOptions = SFSTLExportOptions.Builder()
      .outputPath(outputPath)
      .format(if (options.getString("format") == "ascii") SFSTLFormat.ASCII else SFSTLFormat.BINARY)
      .unit(SFUnit.MILLIMETERS)
      .scale(options.getDouble("scale").toFloat())
      .build()

    session.exportSTL(exportOptions, object : SFExportCallback {
      override fun onSuccess(result: SFSTLResult) {
        val resultMap = Arguments.createMap().apply {
          putString("path",           result.path)
          putInt("fileSize",          result.fileSize.toInt())
          putInt("triangleCount",     result.triangleCount)
          putBoolean("isWatertight",  result.isWatertight)
          putString("checksum",       result.sha256Checksum)
        }
        promise.resolve(resultMap)
      }

      override fun onError(error: SFException) {
        promise.reject("SF_EXPORT_ERROR", "STL-Export fehlgeschlagen: ${error.message}", error)
      }
    })
  }

  // ─────────────────────────────────────────────────────────────────────────
  // Session beenden
  // ─────────────────────────────────────────────────────────────────────────

  @ReactMethod
  fun endSession(sessionId: String, promise: Promise) {
    reactContext.runOnUiQueueThread {
      activeSessions[sessionId]?.cancel()
      activeSessions.remove(sessionId)

      arControllers[sessionId]?.let { arVC ->
        arVC.stop()
        arVC.getView(reactContext).also { view ->
          (view.parent as? ViewGroup)?.removeView(view)
        }
      }
      arControllers.remove(sessionId)

      val result = Arguments.createMap().apply { putString("status", "ended") }
      promise.resolve(result)
    }
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun getVersion(): String = SnugFitManager.getInstance().sdkVersion

  // ─────────────────────────────────────────────────────────────────────────
  // Event-Helper
  // ─────────────────────────────────────────────────────────────────────────

  private fun sendEvent(eventName: String, params: WritableMap) {
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(eventName, params)
  }

  private fun parseResolution(value: String?): SFScanResolution = when (value) {
    "ultra"    -> SFScanResolution.ULTRA
    "standard" -> SFScanResolution.STANDARD
    else       -> SFScanResolution.HIGH
  }
}

File: snugfit-integration/ios/SnugFitBridge.swift
/**
 * ─────────────────────────────────────────────────────────────────────────────
 * SnugFitBridge.swift  –  iOS Native Module für React Native
 * ─────────────────────────────────────────────────────────────────────────────
 * Verbindet das React Native JavaScript-Layer mit dem nativen SnugFit SDK.
 *
 * Setup:
 *   1. SnugFit.xcframework zum Xcode-Projekt hinzufügen (General → Frameworks)
 *   2. Diese Datei ins ios/-Verzeichnis des RN-Projekts kopieren
 *   3. Kamera-Berechtigungen in Info.plist:
 *      NSCameraUsageDescription = "Wird für den 3D-Fußscan benötigt"
 *
 * Benötigt iOS 14+ (LiDAR), iOS 13+ (ohne LiDAR/ARKit Depth)
 * ─────────────────────────────────────────────────────────────────────────────
 */

import Foundation
import React
import SnugFitSDK          // ← offizielles SnugFit Framework (Pseudocode)
import ARKit
import SceneKit

@objc(SnugFitSDK)
class SnugFitBridge: RCTEventEmitter {

  // ── Interne Zustandsvariablen ──────────────────────────────────────────────
  private var activeSessions:    [String: SFScanSession]    = [:]
  private var arControllers:     [String: SFARViewController] = [:]
  private var positioningTimers: [String: Timer]             = [:]

  // ── RCTEventEmitter Setup ─────────────────────────────────────────────────

  override func supportedEvents() -> [String] {
    return [
      "SnugFitPositionUpdate",    // Echtzeit-Positions-Feedback
      "SnugFitMeshProgress",      // Verarbeitungsfortschritt 0–100%
      "SnugFitScanQuality",       // Scan-Qualitäts-Update
    ]
  }

  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - SDK Initialisierung
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Initialisiert das SnugFit SDK mit API-Key und Konfiguration.
   * Muss einmalig beim App-Start aufgerufen werden.
   *
   * PSEUDOCODE: `SnugFitManager.configure()` ist der echte SDK-Einstiegspunkt.
   * Namen können je nach SDK-Version abweichen.
   */
  @objc func initialize(
    _ config: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      do {
        // PSEUDOCODE ─ SnugFit SDK konfigurieren
        let sdkConfig = SFConfiguration(
          apiKey:      config["apiKey"]      as? String ?? "",
          environment: config["environment"] as? String == "sandbox" ? .sandbox : .production,
          resolution:  self.parseResolution(config["resolution"] as? String),
          meshModel:   config["meshModel"]   as? String ?? "foot_v3",
          enableLiDAR: config["enableLiDAR"] as? Bool ?? true,
          enableARKit:  config["enableARMode"] as? Bool ?? true
        )

        // PSEUDOCODE ─ Initialisierung aufrufen
        try SnugFitManager.shared.configure(with: sdkConfig)

        resolve([
          "version":     SnugFitManager.shared.sdkVersion,
          "lidarAvailable": ARWorldTrackingConfiguration.supportsSceneReconstruction(.mesh),
          "status":      "initialized",
        ])
      } catch {
        reject("SF_INIT_ERROR", "SDK-Initialisierung fehlgeschlagen: \(error.localizedDescription)", error)
      }
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - Session Management
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Erstellt eine neue Scan-Session für einen Fuß.
   */
  @objc func createScanSession(
    _ options: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    guard let footSideStr = options["footSide"] as? String,
          let footSide    = SFFootSide(rawValue: footSideStr) else {
      reject("SF_INVALID_PARAMS", "footSide muss 'LEFT' oder 'RIGHT' sein", nil)
      return
    }

    // PSEUDOCODE ─ Session erstellen
    let sessionOptions = SFSessionOptions(
      footSide:        footSide,
      captureTimeout:  options["captureTimeout"] as? TimeInterval ?? 45,
      minQuality:      options["minQuality"] as? Float ?? 92.0,
      mode:            options["mode"] as? String == "freeform" ? .freeform : .guided
    )

    // PSEUDOCODE ─ Session-Objekt vom SDK anfordern
    let session   = SnugFitManager.shared.createSession(with: sessionOptions)
    let sessionId = session.sessionId

    activeSessions[sessionId] = session
    resolve(["sessionId": sessionId])
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - AR-Kamera-Preview
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Startet die AR-Kamera-Preview und bindet das SnugFit AR-Overlay
   * an den nativen View des React Native-Rendering-Tree.
   */
  @objc func startARPreview(
    _ options: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      guard let sessionId = options["sessionId"]  as? String,
            let session   = self.activeSessions[sessionId],
            let viewHandle = options["viewHandle"] as? NSNumber,
            let bridge    = self.bridge,
            let uiManager = bridge.uiManager,
            let view      = uiManager.view(forReactTag: viewHandle) else {
        reject("SF_PREVIEW_ERROR", "Session oder View nicht gefunden", nil)
        return
      }

      // PSEUDOCODE ─ AR-View-Controller erstellen und an View binden
      let arConfig = SFARConfig(
        showDepthMap: options["showDepthMap"] as? Bool ?? true,
        showSkeleton: options["showSkeleton"] as? Bool ?? true,
        overlayColor: UIColor.systemGreen
      )

      let arVC = SnugFitManager.shared.createARViewController(
        for:    session,
        config: arConfig
      )

      // AR-View in den React Native View einbetten
      arVC.view.frame = view.bounds
      arVC.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
      view.addSubview(arVC.view)

      // Position-Updates vom AR-Controller abonnieren
      arVC.onPositionUpdate = { [weak self] positionData in
        guard let self = self else { return }
        self.sendEvent(
          withName: "SnugFitPositionUpdate",
          body: [
            "sessionId":  sessionId,
            "status":     positionData.statusString,
            "confidence": positionData.confidence,
            "boundingBox": [
              "x":      positionData.boundingBox.origin.x,
              "y":      positionData.boundingBox.origin.y,
              "width":  positionData.boundingBox.size.width,
              "height": positionData.boundingBox.size.height,
            ]
          ]
        )
      }

      self.arControllers[sessionId] = arVC
      resolve(["status": "ar_preview_started"])
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - Winkel-Aufnahme
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Nimmt einen Scan-Winkel auf.
   */
  @objc func captureAngle(
    _ options: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    guard let sessionId = options["sessionId"] as? String,
          let session   = activeSessions[sessionId],
          let angleId   = options["angleId"] as? String else {
      reject("SF_CAPTURE_ERROR", "Ungültige Parameter", nil)
      return
    }

    // PSEUDOCODE ─ Winkel-Aufnahme
    let captureOptions = SFCaptureOptions(
      angleId:         angleId,
      frameCount:      options["frameCount"] as? Int ?? 5,
      depthResolution: .full
    )

    session.captureAngle(captureOptions) { result in
      switch result {
      case .success(let captureResult):
        resolve([
          "success":  true,
          "quality":  captureResult.quality,
          "message":  "Aufnahme erfolgreich",
          "previewImageBase64": captureResult.previewImageBase64 ?? NSNull(),
        ])

      case .failure(let error):
        resolve([
          "success":  false,
          "quality":  error.quality ?? 0,
          "message":  error.localizedDescription,
        ])
      }
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - Mesh-Verarbeitung
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Verarbeitet alle aufgenommenen Winkel zu einem 3D-Mesh.
   */
  @objc func processMesh(
    _ options: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    guard let sessionId = options["sessionId"] as? String,
          let session   = activeSessions[sessionId] else {
      reject("SF_MESH_ERROR", "Session nicht gefunden", nil)
      return
    }

    let meshOptions = SFMeshOptions(
      targetTriangles:  options["targetTriangles"] as? Int ?? 50_000,
      smoothing:        options["smoothing"]       as? Float ?? 0.3,
      fillHoles:        options["fillHoles"]       as? Bool  ?? true,
      coordinateSystem: .footStandard
    )

    // Progress-Events feuern
    session.processMesh(options: meshOptions, progress: { [weak self] percentage in
      guard let self = self else { return }
      self.sendEvent(
        withName: "SnugFitMeshProgress",
        body: ["sessionId": sessionId, "percentage": percentage]
      )
    }, completion: { result in
      switch result {
      case .success(let meshResult):
        resolve([
          "accuracy": meshResult.accuracy,
          "meshId":   meshResult.meshId,
          "measurements": [
            "length":       meshResult.measurements.length,
            "width":        meshResult.measurements.width,
            "heelWidth":    meshResult.measurements.heelWidth,
            "instepHeight": meshResult.measurements.instepHeight,
            "archHeight":   meshResult.measurements.archHeight,
            "ballGirth":    meshResult.measurements.ballGirth,
          ],
        ])

      case .failure(let error):
        reject("SF_MESH_ERROR", "Mesh-Verarbeitung fehlgeschlagen: \(error.localizedDescription)", error)
      }
    })
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - STL-Export
  // ─────────────────────────────────────────────────────────────────────────

  /**
   * Exportiert das Mesh als STL-Datei.
   */
  @objc func exportSTL(
    _ options: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    guard let sessionId  = options["sessionId"]  as? String,
          let session    = activeSessions[sessionId],
          let outputPath = options["outputPath"] as? String else {
      reject("SF_EXPORT_ERROR", "Ungültige Export-Parameter", nil)
      return
    }

    let exportOptions = SFSTLExportOptions(
      outputPath:  outputPath,
      format:      options["format"] as? String == "ascii" ? .ascii : .binary,
      unit:        .millimeters,
      scale:       options["scale"] as? Double ?? 1.0,
      metadata:    options["metadata"] as? [String: String] ?? [:]
    )

    session.exportSTL(options: exportOptions) { result in
      switch result {
      case .success(let stlResult):
        resolve([
          "path":          stlResult.path,
          "fileSize":      stlResult.fileSize,
          "triangleCount": stlResult.triangleCount,
          "isWatertight":  stlResult.isWatertight,
          "checksum":      stlResult.sha256Checksum,
        ])

      case .failure(let error):
        reject("SF_EXPORT_ERROR", "STL-Export fehlgeschlagen: \(error.localizedDescription)", error)
      }
    }
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - Session beenden
  // ─────────────────────────────────────────────────────────────────────────

  @objc func endSession(
    _ sessionId: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject:  @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      if let session = self.activeSessions[sessionId] {
        session.cancel()
        self.activeSessions.removeValue(forKey: sessionId)
      }
      if let arVC = self.arControllers[sessionId] {
        arVC.view.removeFromSuperview()
        self.arControllers.removeValue(forKey: sessionId)
      }
      resolve(["status": "ended"])
    }
  }

  @objc func getVersion() -> String {
    // PSEUDOCODE ─ SDK-Version
    return SnugFitManager.shared.sdkVersion
  }

  // ─────────────────────────────────────────────────────────────────────────
  // MARK: - Hilfsfunktionen
  // ─────────────────────────────────────────────────────────────────────────

  private func parseResolution(_ value: String?) -> SFScanResolution {
    switch value {
    case "ultra":    return .ultra
    case "standard": return .standard
    default:         return .high
    }
  }
}
